### DISCUSSION TITLE
Actual state of rewriting BleskOS to C

### DISCUSSION THREAD

**Discussion by VendelinSlezak** [VendelinSlezak/BleskOS#16] [Announcements]
BleskOS was succesfully rewritten.

**Comment by VendelinSlezak** [comment]
graphic_vesa.asm and graphic_font.asm are successfully rewritten. I've add many improvements, two of most important are changeable size of pen and drawing of quadratic bezier curves. This is small preview:

![VirtualBox_BleskOS 2022_12_29 system_graphic](https://user-images.githubusercontent.com/43180618/209946517-3fa76c8b-9417-4e27-af5d-d153defe4207.png)

**Comment by VendelinSlezak** [comment]
system_memory_management.asm is successfully rewritten. Now we can use methods malloc(), aligned_malloc(), calloc(), realloc() and free(). That allow much better memory management than one in system_memory_management.asm

**Comment by VendelinSlezak** [comment]
system_interrupts.asm/system_pci.asm/system_time.asm/system_wait.asm are successfully rewritten.

**Comment by NDRAEY** [comment]
Wow

**Comment by ghost** [comment]
Seeing as how fast you've done rewritting those scripts, it will probably be done faster

**Comment by VendelinSlezak** [comment]
I hope so. So far it goes much faster than I thought it would be. I today also successfully rewritten system_acpi.asm, hardware_ps2_controller.asm, hardware_ps2_keyboard.asm and hardware_ps2_mouse.asm.

**Comment by ghost** [comment]
Will you ever try simple 3D graphics, such as a wireframe spinning cube? Or making a simple topdown game?

**Comment by VendelinSlezak** [comment]
Yes, 3d graphic is also on my list, however it do not have high priority now. Probably during this year I will try to implement some very simple 3d library to draw wireframed objects. 

hardware_ide.asm/hardware_paralel_ata.asm/hardware_paralel_atapi.asm are successfully rewritten.

**Comment by Aap-Dev** [comment]
Looks awesome like everything you do👍
Any chance you will improve the web browser?

**Comment by VendelinSlezak** [comment]
Yes, it will be one of biggest updates with this rewriting. I am planning to add basic CSS, what will very much improve looking of pages. Actually, about month ago, I already wrote some code in assembly about this, so I will just finalize it.
Today I rewritten system_ac97.asm and most of system_hda.asm.

**Comment by VendelinSlezak** [comment]
I today rewritten everything what was in system_hda.asm and I was also improving this driver. I hope I will be soon able to release version that will also play sound on real hardware, not only in emulator. Today version was on real hardware able to parse right path through nodes, and also running stream, so it must be very close to playing sound. I am also planning to implement microphone input.

**Comment by VendelinSlezak** [comment]
I rewritten hardware_ahci.asm and hardware_serial_ata.asm. As holiday ends and school begins, I cannot afford investing very much time to BleskOS, but I will try to find enough time to continue rewriting.

**Comment by VendelinSlezak** [comment]
I successfully rewritten network_ethernet_nic_intel.asm

**Comment by VendelinSlezak** [comment]
I successfully rewritten network_ethernet_nic_amd.asm

**Comment by VendelinSlezak** [comment]
I successfully written driver for serial cdrom, I started writing this driver from time I wrote driver for serial hdd(hardware_serial_ata.asm), but now it finally works.

**Comment by VendelinSlezak** [comment]
I wrote driver for realtek 8169 compatibile ethernet cards. This is big step for BleskOS, because about half of computes today have some realtek 8169 compatibile card. Of course I also tested that this driver works on real computers. I also started to rewriting network_protocols.asm. So far I rewritten ethernet protocol, IP protocol, UDP protocol, DHCP protocol, DNS protocol and I am working on TCP protocol.

**Comment by VendelinSlezak** [comment]
I make first version of BleskOS scalable font, so far with characters from A to Z. As you can see, font is so far not very nice, and there are some minor rendering mistakes, but overall, it is good start.

![VirtualBox_BleskOS 2022_31_01_2023_18_23_00](https://user-images.githubusercontent.com/43180618/215836971-286d2266-df2c-491d-9fe3-df8bba246e1c.png)

**Comment by VendelinSlezak** [comment]
I rewritten almost everything from network_protocols.asm and network_file_transfer.asm. I also add characters from a to f to scalable font.

**Comment by VendelinSlezak** [comment]
I finailzed rewriting of network_protocols.asm and network_file_transfer.asm and I also started rewritting of filesystem_iso9660.asm and hardware_storage_center.asm

**Comment by VendelinSlezak** [comment]
I rewritten hardware_storage_center.asm. From now it will be very easy to write drivers for any filesystem and insert them into existing code. That was quite problem in my previous assembly code, because I was using hardwired hard disk -> JUS, cdrom -> ISO9660, usb flash -> FAT32. now it is not only possible to easy write driver for any filesystem, but I also add support for multiple partitions. So basically, if you will have installed BleskOS on computer with other operation system, you will be able to access data of that operation system on another parition. I also deciced to abandon JUS and just use FAT instead. JUS was my filesystem and it was designed really badly. So I decided to completely abandon it and just use what is verified by years.

**Comment by VendelinSlezak** [comment]
I successfully rewritten filesystem_iso9660.asm and filesystem_fat.asm. FAT driver have many great improvements, most important are full support for long file names and that it now works not only with FAT32, but also with FAT16 and FAT12. So far from folder drivers is rewritten about 80-85% code, last thing are USB. Now I have holidays, so I will be able to put more time to BleskOS, so I hope I will continue faster.

**Comment by VendelinSlezak** [comment]
Unfortunately, in the end I did not have very much time last weeks, I only rewritten usb_center.asm and usb_uhci.asm.

**Comment by VendelinSlezak** [comment]
I rewritten usb_ohci.asm, usb_mouse.asm, usb_keyboard.asm and hardware_keyboard_layouts.asm. I am adding many improvements. I also fixed problem with receiving packets from mouse and keyboard on real hardware that was happening in my assembly code.

**Comment by VendelinSlezak** [comment]
I wrote parser of HID devices, so now under BleskOS work almost any USB mouse.

**Comment by VendelinSlezak** [comment]
After final debugging of code, I tested that every hardware I have access to - 3 USB keyboards and 8 USB mouses including one wireless mouse - works under BleskOS.

**Comment by VendelinSlezak** [comment]
I successfully rewritten usb_ehci.asm and usb_mass_storage.asm. USB stack has many improvements that actual version is not even comparable with assembly version. I will mention just a few:
- parser for HID devices
- automatic detection of USB devices
- working code for USB keyboards(assembly version was not really working even in emulator, not even on real hardware)
- reading/writing speed to USB mass storage device improved by allowing to transfer more than 1 sector per transfer
- USB mass storage devices can be acessed also through UHCI and OHCI controllers, not only through EHCI

This means that everything from folder drivers is now rewritten. I will took some time to finalize my rewritten code, and than I will start rewriting programs. Folder gui will be rewritten alongside with programs.

**Comment by VendelinSlezak** [comment]
I successfully rewritten library_log.asm. For example in QEMU with CD disk and USB mouse attached output of logs looks like this:
![Snímka obrazovky_2023-04-24_17-01-49](https://user-images.githubusercontent.com/43180618/234037199-0e799b94-ad62-4ffc-bfe8-33e9166a7c82.png)
I also started writing File dialog. So far it looks like this:
![Snímka obrazovky_2023-04-24_17-02-28](https://user-images.githubusercontent.com/43180618/234037316-e316d898-6a80-45d6-bfd0-e9122703d29a.png)

**Comment by VendelinSlezak** [comment]
I wrote decoder for compression algorithm DEFLATE. This is great step forward, because DEFLATE is used in formats such as PNG, ZIP, PDF, DOCX, XLSX, PPTX, ODT, ODS, ODP. So we are step closer to have these formats in BleskOS.

**Comment by VendelinSlezak** [comment]
I apologize for recent silence. I was not working on BleskOS at all. Fortunately, I now have time to continue rewriting so I hope I will be able to rewrite BleskOS as soon as possible. I wrote main window. It is improved and it look like this:
![Snímka obrazovky_2023-06-03_18-25-26](https://github.com/Klaykap/BleskOS/assets/43180618/a4da1266-6aa1-4833-b36b-b728a2628c4a)
I started rewriting text editor. I add menu, so user do not need use only keyboard for file dialogue.
![Snímka obrazovky_2023-06-03_18-25-40](https://github.com/Klaykap/BleskOS/assets/43180618/278ff9bc-ce11-4bb7-9856-27f51643f520)
And very nice improvement is that now it is possible to have opened more files at same time in one program.
![Snímka obrazovky_2023-06-03_18-27-00](https://github.com/Klaykap/BleskOS/assets/43180618/7d3d9510-3331-4d7a-80f5-f7e73d81e2da)
I also finish Save file dialog, so that mean that library_file_dialogue.asm, board_main.asm and board_hardware_centrum.asm are rewritten.

**Comment by VendelinSlezak** [comment]
Program Text editor is fully rewritten with few improvements.

**Comment by VendelinSlezak** [comment]
I rewritten Media viewer for images. There are great improvements, you can now zoom images, and I wrote code for decoding PNG files, so BleskOS can already open almost all PNG images.
![Snímka obrazovky_2023-07-17_14-39-00](https://github.com/Klaykap/BleskOS/assets/43180618/d30b8aeb-b33e-4887-a9f7-75a15eda5d65)

**Comment by VendelinSlezak** [comment]
I succesfully rewritten Graphic editor. Now you can change pen size and zoom image. And I also written code for decoding GIF images, so BleskOS can now also open almost all GIF images. However, animated GIFs are not supported now, from animated GIF will BleskOS show only first image, but I am planning to write code also for them.
![Snímka obrazovky_2023-07-20_12-57-21](https://github.com/Klaykap/BleskOS/assets/43180618/80265c6f-7c98-40e9-9ea6-265f958c9671)

**Comment by VendelinSlezak** [comment]
I fully rewritten Media Viewer. Here is how actual version look like:
![Snímka obrazovky_2023-07-25_17-53-18](https://github.com/Klaykap/BleskOS/assets/43180618/cee23370-ee80-49c8-afe8-027578095a80)
![Snímka obrazovky_2023-07-27_18-10-51](https://github.com/Klaykap/BleskOS/assets/43180618/498aaabe-8413-47b4-8110-6da2cc49b009)
![Snímka obrazovky_2023-07-25_18-14-58](https://github.com/Klaykap/BleskOS/assets/43180618/3bb9d6a7-547b-4b55-9823-796fd6c53b9e)

**Comment by VendelinSlezak** [comment]
BleskOS scalable font now have all ASCII chars:
![Snímka obrazovky_2023-07-29_17-29-23](https://github.com/Klaykap/BleskOS/assets/43180618/4eb57875-f535-452b-a7cd-9c1176a60d99)

**Comment by VendelinSlezak** [comment]
I started rewriting Internet Browser. So far it can not acess online pages, I am only showing HTML files. As example page I will use wiki.osdev.org Here is how it looks like now:
![Snímka obrazovky_2023-08-03_19-19-20](https://github.com/Klaykap/BleskOS/assets/43180618/245f35ee-8f3b-4827-aaba-12e86b68bce1)
As you can see, it looks similar to previous version of Internet browser, but it is better, thanks to scalable font. But in reality it is even better, because actual code is ready to support CSS styles, I only have to write parser for it. Now it shows page with default CSS styles for tags. I hope I will soon have version with basic CSS support that will look even much better than this.

**Comment by VendelinSlezak** [comment]
First basic CSS parser brings Internet browser to whole new level:
![Snímka obrazovky_2023-08-05_13-03-47](https://github.com/Klaykap/BleskOS/assets/43180618/7265b9cb-9d66-4731-b0c6-8016fb147ea0)

**Comment by VendelinSlezak** [comment]
Now page wiki.osdev.org look almost exactly like in any normal browser(keep in mind that we are displaying only HTML file, without CSS files that are downloaded when you open page online):
![Snímka obrazovky_2023-08-08_18-08-57](https://github.com/Klaykap/BleskOS/assets/43180618/749706dc-ae9f-497e-8800-53bcf8f4fffa)

**Comment by VendelinSlezak** [comment]
Finally, I sucessfully for the first time connected BleskOS to network on real computer.
![80A32151-C53D-41E5-8AD8-28F4CFDBE107](https://github.com/Klaykap/BleskOS/assets/43180618/db425694-ab03-44d0-9e33-382c61ec35be)

**Comment by TheEccentricGenius** [comment]
How much longer do you have?

And will this support WiFi?

**Comment by VendelinSlezak** [comment]
I rewrote Internet browser to version that can be released. I decided that I will not rewrite Document editor, Table editor and File manager yet, but only after I will release actual code. So now I am carefully reading all code I rewrote, to catch all bugs I will be able to found. I do not know how long it will exactly took, but I will post how many of code is reviewed. After review, I will update repository. So keep watching.

@Wifi: I would love to have drivers for Wifi, but I found that it is incredibly hard to obtain any documentation on this subject or some wireless card specification. In fact, so far I do not have any except for some notes on https://wiki.osdev.org/Intel_Dual_Band_Wireless and knowing that there are some open-source drivers. So unfortunately, in the near future BleskOS will not have Wifi support. However, I have it on my list, so if I will found way to write driver, I will surely do.

**Comment by VendelinSlezak** [comment]
I greatly improved VBE driver. Previous version was not able to use everything what VBE gives, so it was capable of setting best graphic mode only on 1 computer from my 10 testing computers. Improved version use all VBE graphic modes and is able to set best mode on 5 computers + on one is mode not best, but it is in same ration as best mode, so it is not blurry.

**Comment by VendelinSlezak** [comment]
I went through about two third of drivers and I improved tons of things and fixed many bugs. Released version will not be just rewritten old BleskOS plus some new features, it will be code on completely new level.
Just to mention one improvement. Actual code in repository is able to read from USB flash around 1 MB per second. Rewritten code was able to achieve similar performance. Now I rewrote code and it can achieve around 5 MB per second.

**Comment by VendelinSlezak** [comment]
If everything goes well, I will release BleskOS 2023 update 1 this week.

**Reply by NDRAEY** [reply]
🚀

**Reply by davidgaosheng** [reply]
you are doing a wonderful job. thanks

**Comment by VendelinSlezak** [comment]
Now all code is rewritten. I will give it some final touches and write better README. So from actual state I think I will release BleskOS tomorrow. And two important things. BleskOS will be released under MIT license. And here is how new description of BleskOS look like:

> BleskOS is 32-bit operating system written from scratch. One of main features is alternative graphic user interface that do not use windows, as classic operating systems do, but switches between programs on full screen. BleskOS is designed for older computers. Operating systems of today have very high hardware requirements that make computers even 10 years old completly unusable. Goal of BleskOS is to provide system, that could be installed on such computers and make them usable again.

I did not reread all code I rewritten, because everything of this already took too much time and I am sure you all will like it more, when you will see improvements along the way. So not everything is perfect and I see quite a lot of thing I want to improve, but actual version is usable, so I will release it and from this point you will see BleskOS development online.

**Comment by VendelinSlezak** [comment]
Repository is updated.

**Reply by davidgaosheng** [reply]
hi, congratulations! we finally see your C programs.

**Reply by davidgaosheng** [reply]
I remembered that you mentioned your email address, but sorry I forgot. would you please tell me once more. thanks

**Reply by VendelinSlezak** [reply]
klaykap()yandex.com (replace () with @)
